mediagrains_async/__init__.py

```diff
@@ -0,0 +1,548 @@
+                meta['grain']['cog_coded_audio']['remainder'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_audio']['sample_rate'] = await gbhd_child.read_uint(4)
+
+            elif gbhd_child.tag == "eghd":
+                meta['grain']['grain_type'] = "event"
+            else:
+                raise GSFDecodeError(
+                    "Unknown type {} at offset {}".format(gbhd_child.tag, gbhd_child.block_start),
+                    gbhd_child.block_start,
+                    length=gbhd_child.size
+                )
+
+        return meta
+
+    async def decode_file_headers(self):
+        """Verify the file is a supported version, and get the file header
+
+        :returns: File header data (segments and tags) as a dict
+        :raises GSFDecodeBadVersionError: If the file version is not supported
+        :raises GSFDecodeBadFileTypeError: If this isn't a GSF file
+        :raises GSFDecodeError: If the file doesn't have a "head" block
+        """
+        (major, minor) = await self._decode_ssb_header()
+        if (major, minor) != (7, 0):
+            raise GSFDecodeBadVersionError("Unknown Version {}.{}".format(major, minor), 0, major, minor)
+
+        try:
+            async with AsyncGSFBlock(self.file_data, want_tag="head") as head_block:
+                self.head = await self._decode_head(head_block)
+                return self.head
+        except EOFError:
+            raise GSFDecodeError("No head block found in file", self.file_data.tell())
+
+    async def __aenter__(self):
+        self.start_loc = await self.file_data.tell()
+        await self.decode_file_headers()
+        return self
+
+    async def __aexit__(self, *args, **kwargs):
+        self.head = None
+        await self.file_data.seek(self.start_loc)
+        self.start_loc = None
+
+    def __aiter__(self):
+        return self.grains()
+
+    async def grains(self, local_ids=None, load_lazily=True):
+        """Asynchronous generator to get grains from the GSF file. Skips blocks which aren't "grai".
+
+        The file_data will be positioned after the `grai` block.
+
+        :param local_ids: A list of local-ids to include in the output. If None (the default) then all local-ids will be
+                          included
+        :param skip_data: If True, grain data blocks will be seeked over and only grain headers will be read
+        :param load_lazily: If True, the grains returned will be designed to lazily load data from the underlying stream
+                            only when it is needed. These grain data elements will have an extra 'load' coroutine for
+                            triggering this load, and accessing data in their data element without first awaiting this
+                            coroutine will raise an exception.
+        :yields: (Grain, local_id) tuple for each grain
+        :raises GSFDecodeError: If grain is invalid (e.g. no "gbhd" child)
+        """
+        if self.head is None:
```

Comment:
Change to just wait for `decode_file_headers` if you agree with the comment above regarding `decode_file_headers`.